ublox_derive/src/output.rs

```diff
@@ -29,6 +29,34 @@ fn generate_debug_impl(pack_name: &str, ref_name: &Ident, pack_descr: &PackDesc)
     }
 }
 
+fn generate_serialize_impl(
```

Comment:
Like I mentioned in another comment, you may be able to avoid having to impl serialize for the iterator types (and also avoid the refcell), but having the message serialize impls handle looping thru the iterator. 

```
diff --git a/ublox_derive/src/output.rs b/ublox_derive/src/output.rs
index b97416d..dc19502 100644
--- a/ublox_derive/src/output.rs
+++ b/ublox_derive/src/output.rs
@@ -38,9 +38,18 @@ fn generate_serialize_impl(
     for field in pack_descr.fields.iter() {
         let field_name = &field.name;
         let field_accessor = field.intermediate_field_name();
-        fields.push(quote! {
-            state.serialize_entry(stringify!(#field_name), &self.#field_accessor())?;
-        });
+        if field.is_iter_field() {
+            fields.push(quote! {
+                state.serialize_key(stringify!(#field_name))?;
+                for item in self.#field_accessor() {
+                    state.serialize_value(&item)?;
+                }
+            });
+        } else {
+            fields.push(quote! {
+                state.serialize_entry(stringify!(#field_name), &self.#field_accessor())?;
+            });
+        }
     }
 
     quote! {
diff --git a/ublox_derive/src/types.rs b/ublox_derive/src/types.rs
index f9b9c9a..46b3cfc 100644
--- a/ublox_derive/src/types.rs
+++ b/ublox_derive/src/types.rs
@@ -149,6 +149,17 @@ impl PackField {
             false
         }
     }
+    pub fn is_iter_field(&self) -> bool {
+        if !self.is_field_raw_ty_byte_array() || !self.has_intermediate_type() {
+            return false;
+        }
+        if let syn::Type::Path(ref p) = self.intermediate_type() {
+            let last = p.path.segments.last().unwrap().to_token_stream().to_string();
+            last.ends_with("Iter")
+        } else {
+            false
+        }
+    }
 }
 
 pub struct UbxExtendEnum {
```

`is_iter_field` is a little jank, not sure the best way to figure out "is this field an iterator field", but it seems to work

Reply:
i had the exact same idea but wasn't sure of a better way to type check if its an iterator or not, didn't want to match based on 'Iter' cuz that was super wonky, was thinkin to make like
`[ubx(iterator)]`
or something that would generate that given that attribute

Reply:
if we go for some attribute then we can customize the serde implementation / autogenerate whether we want based on the field or something ? maybe better / more options easier to change if we want naming or just bits

(then we'd need like 2 different attributes, one for normal enum based and one for function based i.e. `fn has_pr_prr_correction() -> bool`)

Reply:
Yeah I think adding `#[ubx_iterator]` type thing is probably the most robust way to do it - there really isn't a good way to get type information in a macro like this

Reply:
also edit: debug inside `serde::Serialize` would give:
    "fix_stat": "FixStatusInfo { has_pr_prr_correction: false, map_matching: None }",
whereas normally looks like 
```
    "fix_stat": {
      "has_pr_prr_correction": false,
      "map_matching": "None"
    },
```
so maybe cant just use `format!("{:?}")` in the case we want named serialization

Reply:
Also I will say the string checking isn't toooo bad in this case. This macro only has one consumer - the ublox crate. So it wouldn't be unreasonable to enforce a convention in this crate that field iterators have to end in `Iter`. If this macro was meant to be used as a library, like serde or something, that would be a different story. So could go either way on this one :shrug:

Reply:
tbh i just brought that up so i can have an excuse to practice creating some attribute in case manually creating serde impl is not sufficient - (debug is manually implemented anyways so no big deal in the end)